### ISSUE TITLE
BigQuery: Unparsable section when SELECT is wrapped in ()

### ISSUE THREAD

**Issue by tuzonghua** [sqlfluff/sqlfluff#3960]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Trying to lint a BigQuery SQL file produced a parsing error. Removing the `()` around the two SELECT statements fixed the parsing error, but results in invalid BigQuery code (BQ error: `Syntax error: Different set operations cannot be used in the same query without using parentheses for grouping at [16:1]`).

### Expected Behaviour

Run sqlfluff on the SQL code without encountering a parsing error.

### Observed Behaviour

Output when SELECT statements enclosed in `()` [valid BigQuery code]:
```
$ sqlfluff parse test.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        unparsable:                                           !! Expected: 'statement'
[L:  1, P:  1]      |            raw:                                              'WITH'
[L:  1, P:  5]      |            whitespace:                                       ' '
[L:  1, P:  6]      |            raw:                                              'filenames'
[L:  1, P: 15]      |            whitespace:                                       ' '
[L:  1, P: 16]      |            raw:                                              'AS'
[L:  1, P: 18]      |            whitespace:                                       ' '
. . .
[L: 22, P: 17]      |            raw:                                              'FROM'
[L: 22, P: 21]      |            whitespace:                                       ' '
[L: 22, P: 22]      |            raw:                                              'filenames'
[L: 22, P: 31]      |            newline:                                          '\n'
[L: 23, P:  1]      |            raw:                                              ')'
[L: 23, P:  2]      |    newline:                                                  '\n'
[L: 24, P:  1]      |    [META] end_of_file:

==== parsing violations ====
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: "WITH filenames AS
                       | (\n SELECT 'test1' AS ..."
WARNING: Parsing errors found and dialect is set to 'bigquery'. Have you configured your dialect correctly?
```

Output when SELECT statements are not wrapped in `()` [invalid BigQuery code]:
Altered SQL:
```sql
SELECT filename FROM filenames
EXCEPT DISTINCT
SELECT filename FROM expected


UNION ALL


SELECT filename FROM expected
EXCEPT DISTINCT
SELECT filename FROM filenames
```
Logs:
```
$ sqlfluff parse test.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        with_compound_statement:
[L:  1, P:  1]      |            keyword:                                          'WITH'
[L:  1, P:  5]      |            whitespace:                                       ' '
[L:  1, P:  6]      |            common_table_expression:
[L:  1, P:  6]      |                naked_identifier:                             'filenames'
[L:  1, P: 15]      |                whitespace:                                   ' '
[L:  1, P: 16]      |                keyword:                                      'AS'
. . .
[L: 22, P: 17]      |                    [META] dedent:
[L: 22, P: 17]      |                    from_clause:
[L: 22, P: 17]      |                        keyword:                              'FROM'
[L: 22, P: 21]      |                        whitespace:                           ' '
[L: 22, P: 22]      |                        from_expression:
[L: 22, P: 22]      |                            [META] indent:
[L: 22, P: 22]      |                            from_expression_element:
[L: 22, P: 22]      |                                table_expression:
[L: 22, P: 22]      |                                    table_reference:
[L: 22, P: 22]      |                                        naked_identifier:     'filenames'
[L: 22, P: 31]      |                            [META] dedent:
[L: 22, P: 31]      |    newline:                                                  '\n'
[L: 23, P:  1]      |    newline:                                                  '\n'
[L: 24, P:  1]      |    [META] end_of_file:
```

### How to reproduce

Test SQL:
```sql
WITH filenames AS (
  SELECT 'test1' AS filename UNION ALL
  SELECT 'test2' AS filename
),

expected AS (
  SELECT '00' || LPAD(CAST(num AS string), 2, '0') || '_part_00.parquet' AS filename
  FROM UNNEST(GENERATE_ARRAY(0, 31)) AS num
)

(
SELECT filename FROM filenames
EXCEPT DISTINCT
SELECT filename FROM expected
)

UNION ALL

(
SELECT filename FROM expected
EXCEPT DISTINCT
SELECT filename FROM filenames
)
```

Commands run:
`$ sqlfluff parse test.sql`

### Dialect

BigQuery

### Version

sqlfluff: 1.3.2
Python: 3.10.7

### Configuration

[sqlfluff]
dialect = bigquery
templater = jinja
exclude_rules = L029,L031,L032

[sqlfluff:rules]
max_line_length = 120
tab_space_size = 2
indent_unit = space
capitalization_policy = upper

[sqlfluff:rules:L030]
# Function names
extended_capitalisation_policy = upper

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
